core/shared/src/main/scala/hedgehog/core/Seed.scala

```diff
@@ -38,17 +38,46 @@ case class Seed(seed: MersenneTwister64) {
     // Have updated to use the more stable haskell version
     // http://hackage.haskell.org/package/random-1.1/docs/src/System.Random.html#randomRFloating
     val (s2, next) = seed.nextDouble
-    (Seed(s2), 2.0 * (0.5 * from + next * (0.5 * to - 0.5 * from)))
+    (Seed(s2, source), 2.0 * (0.5 * from + next * (0.5 * to - 0.5 * from)))
   }
 }
 
 object Seed {
 
   // FIX: predef IO
   def fromTime(): Seed =
-    fromLong(System.nanoTime)
+    fromSeedSource(SeedSource.FromTime(System.nanoTime))
 
   def fromLong(seed: Long): Seed =
-    Seed(MersenneTwister64.fromSeed(seed))
+    Seed(MersenneTwister64.fromSeed(seed), SeedSource.FromLong(seed))
+  
+  def fromSeedSource(source: SeedSource): Seed =
+    Seed(MersenneTwister64.fromSeed(source.seed), source)
+
+  def fromEnvOrTime(): Seed = {
+    val source = sys.env
+      .get("HEDGEHOG_SEED")
+      .flatMap(s => scala.util.Try(s.toLong).toOption)
+      .map(SeedSource.FromEnv)
+      .getOrElse(SeedSource.FromTime(System.nanoTime()))
+    Seed.fromSeedSource(source)
+  }
+}
+
+sealed trait SeedSource extends Product with Serializable {
```

Comment:
It might not be that important for this PR but do you think we can have constructor methods in the companion object instead of `extends Product with Serializable`?
It's not just for getting rid of `Product` and `Serializable` though. There's also another issue with it. `extends Product with Serializable` still doesn't solve the problem of exposing data instead of type.

More about the reason can be found at https://2020-hindsight-scala.kevinly.dev/docs/adt#use-constructor-methods

Reply:
I agree with you. I guess I just forgot about it. I will add constructor methods. However, unless the actual constructors are private, I would still keep the `extends Product with Serializable` in there.

Reply:
I have no strong objection to keeping them. Just curious, is there any reason to keep them? I'd like learn it. 🙂